node/follower_node.go

```diff
@@ -0,0 +1,507 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+// Package node is the Algorand node itself, with functions exposed to the frontend
+package node
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/algorand/go-deadlock"
+
+	"github.com/algorand/go-algorand/agreement"
+	"github.com/algorand/go-algorand/catchup"
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data"
+	"github.com/algorand/go-algorand/data/account"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/network"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/rpcs"
+	"github.com/algorand/go-algorand/util/execpool"
+)
+
+// AlgorandFollowerNode implements follower mode/ledger delta APIs and disables participation-related methods
+type AlgorandFollowerNode struct {
+	mu        deadlock.Mutex
+	ctx       context.Context
+	cancelCtx context.CancelFunc
+	config    config.Local
+
+	ledger *data.Ledger
+	net    network.GossipNode
+
+	catchupService           *catchup.Service
+	catchpointCatchupService *catchup.CatchpointCatchupService
+	blockService             *rpcs.BlockService
+
+	rootDir     string
+	genesisID   string
+	genesisHash crypto.Digest
+	devMode     bool // is this node operates in a developer mode ? ( benign agreement, broadcasting transaction generates a new block )
+
+	log logging.Logger
+
+	// syncStatusMu used for locking lastRoundTimestamp and hasSyncedSinceStartup
+	// syncStatusMu added so OnNewBlock wouldn't be blocked by oldKeyDeletionThread during catchup
+	syncStatusMu          deadlock.Mutex
+	lastRoundTimestamp    time.Time
+	hasSyncedSinceStartup bool
```

Comment:
For the mutiex, it seems like not since it's only referenced by `OnNewBlock` and `Status`, neither of which are run in separate goroutines.

The other two fields are part of the node status response.

Reply:
I see. The full node wouldn't "use" them, but they all need to be in the implementation of every node because of the API handlers.

Reply:
Yes, it's essentially a dead code path because those APIs only get registered when you enable follower mode, and if you enable follower mode it won't construct a FullNode.

They're required to be there for the interface. We could just replace them with no-ops in the FullNode implementation.

Reply:
Agreed.